### DISCUSSION TITLE
Lack of middleware folder

### DISCUSSION THREAD

**Discussion by cyriaque-ovaga** [nuxt/nuxt#17726] [Questions] [answered]
Hello, I would like to know if we will no longer use the Middleware on Nuxt3.

Thank you

**Comment by lucpotage** [comment, marked as answer]
Middleware still exists in Nuxt 3 but is now under `server`. Have a look at the documentation: https://v3.nuxtjs.org/docs/directory-structure/server#server-middleware.

**Reply by ruslankonev** [reply]
But, this is only server middleware, not client

**Reply by pi0** [reply]
Indeed. Routing middleware are completely different and not supported by Nuxt 3 anymore (at least so far) (although possible to use with nuxt 2 and bridge. learn more: https://nuxtjs.org/docs/directory-structure/middleware/)

**Reply by mgnrfk** [reply]
What could be an acceptable solution then? 
Watching route changes in app.vue to act like a global middleware?

**Reply by wobsoriano** [reply]
> What could be an acceptable solution then? Watching route changes in app.vue to act like a global middleware?

Check the 2nd one https://github.com/nuxt/framework/discussions/906#discussioncomment-1478117

**Reply by HenrijsS** [reply]
> Indeed. Routing middleware are completely different and not supported by Nuxt 3 anymore (at least so far) (although possible to use with nuxt 2 and bridge. learn more: https://nuxtjs.org/docs/directory-structure/middleware/)

Doesn't Nuxt-Auth use Routing middleware??? Why wasn't this mentioned anywhere? 
[Strapi user handling](https://strapi.nuxtjs.org/usage) is unsupported now?

**Comment by wobsoriano** [comment]
For now what we can do is to create something like `server/middleware/guard.ts` and inside it 

```ts
import type { IncomingMessage, ServerResponse } from 'http';

const isAuthenticated = true;

export default (req: IncomingMessage, res: ServerResponse, next) => {
  if (isAuthenticated && req.url === '/login') {
    res.writeHead(301, { Location: '/' });
    res.end();
  } else if (isAuthenticated || req.url === '/login') {
    next();
  } else {
    res.writeHead(301, { Location: '/login' });
    res.end();
  }
};
```

But this will only work on page refresh and not trigger on client side routing.

For client side, I think we can use the layouts and [global state](https://v3.nuxtjs.org/docs/usage/state) for that:

```html
<!-- layouts/authenticated.vue -->
<script setup lang="ts">
  const session = useState('session')
  if (!session.value) {
    // redirect to login
  }
</script>

<template>
  <div>
    <slot />
  </div>
</template>
```

```ts
// plugins/auth.ts
export default defineNuxtPlugin((nuxt) => {
  const session = useState(
    'session',
    () => nuxt.ssrContext.req.session // or depends where session is coming
  )
})
```

Not sure about the client one, haven't tried.

**Reply by kevhof** [reply]
Still, you can't send sensitive data to the user on client side routing with this method. It's a nice "design" solution but not a safe one as this can be tempered with by a malicious user right ? 

For a safe solution, you could use asyncData and create a secure api for it. It works well when there is a small amount of data to show like a profile page for example.

**Reply by beocaca** [reply]
can it work please? do you have demo!!!?

**Comment by beocaca** [comment]
any solution please

**Comment by psycongaroo** [comment]
So Nuxt3 will not support middleware for client?

**Comment by lucpotage** [comment]
I know they are experimenting the Suspense API but it's still experimental. There may be something around this that would replace the notion of middleware in Nuxt 2.

There are also navigation guards in Vue Router 4 that you may define in setup.

**Comment by wobsoriano** [comment]
https://twitter.com/nuxt_js/status/1486482521844064257?s=21

**Comment by davidmyersdev** [comment]
It's here. https://github.com/nuxt/framework/tree/main/examples/routing/middleware
